### ISSUE TITLE
quadrupole moment calculation

### ISSUE THREAD

**Issue by chrisiacovella** [choderalab/modelforge#381]
We want to be able to explore how different properties in the loss function (related to charge) influence the trained model. 

Current in the loss we can have components of: 
- total_charge
- per_atom_charge
- dipole_moment

quadrupole moment should also be considered here as most of our datasets include this quantity and publications with aimnet2 have used it as part of training. 

Quadrupole moment is a 3x3 matrix.  Each element is computed as: 

$$Q_{ij} = \sum_n q_n (3 r_{n,i} r_{n_j} - r_n^2\delta_{ij})$$

where: 
- $$q_n$$ is the partial charge on atom n
- $$r_{n,i}$$ and $$r_{n,j}$$ refer to the cartesian coordinates of positions (r) of atom n
- $$r_n^2=x^2 + y^2 + z^2$$
- $$\delta_{ij}$$ is the Kronecker delta which is 1 if i==j, otherwise 0

While the calculation of the quadrupole moment itself is pretty straightforward, it took me a bit of time to figure out how to implement the algorithm in pytorch to work with  batches of data (without using any sort of looping).   The algorithm relies on computing each piece on a per atom basis and summing at the end.  

For each batch we have as input :
- positions (r) of shape [n_atoms, 3] (note these will have had their COM shifted to the origin)
- per_atom_charge (q) of shape [n_atoms, 1]
-  atomic_sub_system_indices shape [n_atoms]

So for each atom, we can calculate the outer product to get the $$r_{n,i} r_{n,j}$$ term, which results in a tensor of sized [n_atoms, 3, 3]

```
r_outer_prod = torch.einsum('mk,ml->mkl', r, r)
```

Next we can calculate the r^2 term: (this will be of shape [n_atoms] but we can just adjust the shape when we got to multiply)

```
r_squared = torch.einsum('mk,mk->m', r, r)
```

To get the kronecker delta term, we can just make a bunch of of copies of an identity matrix.  The size of `I_batch` will be [n_atoms, 3,3]:

```
I = torch.eye(3)
I_batch = I.unsqueeze(0).expand(r.shape[0], -1, -1)
```

Compute the term $$(3 r_{n,i} r_{n_j} - r_n^2\delta_{ij})$$: 

```
quad_temp = 3 * r_outer_prod - r_squared.view(-1,1,1) * I_batch
```

Multiply by the charge to get the per atom quadrupole moment contributions: 

```
quadrupole_moment_per_atom = torch.einsum('m...,mkl->mkl', q, quad_tenmp)
```

Finally, we need to use scatter add to sum up the per_atom contributions for each system, using the atomic_subsystem_indices (that is just a 1d tensor that defines which atoms are in each system in the batch): 

Figure out the unique systems in the batch
``` 
mol_id = torch.unique(atomic_subsystem_indices)
```
create an empty output tensor of size [n_systems, 3, 3]
```
output_tensor = torch.zeros(mol_id.shape[0], 3, 3)
```

define the indices for scatter add; we will use the atomic_subsystem indices, expanded to match the size of the `quadrupole_moment_per_atom` tensor 

```
indices = atomic_subsystem_indices.view(-1, 1, 1).expand_as(quadrupole_moment_by_atom)
```

perform scatter add to get the quadrupole moment for each molecule in the batch, which will be of shape [n_atoms, 3, 3]

```
quadrupole_moment = output_tensor.scatter_add(0, indices, quadrupole_moment_per_atom)
```
